File: R/reliability_test.R
# ======================================================================
# Reliability test for the Chat Completions grading pipeline
#
# Runs grade_student() N times for every student submission and writes
# one CSV per student containing N rows — one per run. This allows
# variability in LLM grading to be measured across repeated calls with
# identical inputs.
#
# Output files are written beside the student submission folders:
#   {directory_path}/{folder_name}_grades.csv
# e.g. "R assignments/lab-9_student_high_grades.csv"
#
# Usage:
#   N <- 10          # set number of runs (default 10)
#   source("R/reliability_test.R")
# ======================================================================

# ---- config ----
if (!exists("N"))          N          <- 10L   # override before sourcing
if (!exists("LAB_NUMBER")) LAB_NUMBER <- 9L

# ---- deps ----
if (!"librarian" %in% rownames(installed.packages())) install.packages("librarian")
librarian::shelf(httr2, jsonlite, stringr, readr, fs)
if (file.exists(".env")) dotenv::load_dot_env()

# load shared utilities (safe_num)
source("./R/utils.R")

# ---- load grading functions without triggering main() ----
# sys.source() into a child env prevents the globalenv() guard in
# chat_grading_runner.R from firing, so main() is never called.
runner_env            <- new.env(parent = globalenv())
runner_env$LAB_NUMBER <- LAB_NUMBER
sys.source("R/chat_grading_runner.R", envir = runner_env)

grade_student <- runner_env$grade_student

# ---- paths and column layout ----
directory_path  <- runner_env$directory_path

Q_COUNT         <- runner_env$Q_COUNT
Q_COLS          <- paste0("Q", seq_len(Q_COUNT))
Q_FEEDBACK_COLS <- paste0("Q", seq_len(Q_COUNT), "_feedback")
COL_ORDER       <- c("Run", "Total", "OverallComment", Q_COLS, Q_FEEDBACK_COLS)

# ---- helpers ----

#' Grade one student N times and return a data frame with one row per run
#'
#' Calls \code{grade_student()} \code{n_runs} times for the same submission
#' file, recording the run index alongside all grade and feedback fields.
#' Per-run exceptions are caught and recorded as error rows so a single API
#' failure does not abort the student's batch.
#'
#' @param student_file Character. Path to the student's \code{.qmd} file.
#' @param student_name Character. Display name used in progress messages.
#' @param n_runs Integer. Number of grading runs to perform.
#'
#' @returns A \code{data.frame} with \code{n_runs} rows and columns
#'   \code{Run}, \code{Total}, \code{OverallComment},
#'   \code{Q1}–\code{Q{Q_COUNT}}, and
#'   \code{Q1_feedback}–\code{Q{Q_COUNT}_feedback}.
grade_n_times <- function(student_file, student_name, n_runs) {
  records <- vector("list", n_runs)

  for (i in seq_len(n_runs)) {
    message("  Run ", i, "/", n_runs, " ...")

    records[[i]] <- tryCatch({
      result    <- grade_student(student_file)
      questions <- result[["questions"]]

      r <- list(
        Run            = i,
        Total          = safe_num(result[["total"]]),
        OverallComment = as.character(
          if (!is.null(result[["overall_comment"]])) result[["overall_comment"]] else ""
        )
      )
      for (q in Q_COLS) {
        qinfo    <- questions[[q]]
        r[[q]]   <- safe_num(qinfo[["grade"]])
        r[[paste0(q, "_feedback")]] <- as.character(
          if (!is.null(qinfo[["feedback"]])) qinfo[["feedback"]] else ""
        )
      }
      r

    }, error = function(e) {
      message("    ERROR on run ", i, ": ", conditionMessage(e))
      r <- list(Run            = i,
                Total          = NA_real_,
                OverallComment = paste("Error:", conditionMessage(e)))
      for (q in Q_COLS) {
        r[[q]]                      <- NA_real_
        r[[paste0(q, "_feedback")]] <- NA_character_
      }
      r
    })
  }

  do.call(rbind.data.frame, lapply(records, function(x) {
    miss <- setdiff(COL_ORDER, names(x))
    for (m in miss) x[[m]] <- NA
    x <- x[COL_ORDER]
    x <- lapply(x, function(v) if (is.null(v) || length(v) == 0) NA else v)
    as.data.frame(x, stringsAsFactors = FALSE)
  }))
}

# ---- main ----

#' Run the reliability test across all student submissions
#'
#' Walks \code{directory_path} for student submission subfolders, grades
#' each student \code{N} times via \code{grade_n_times()}, and writes one
#' CSV per student beside the submission folder:
#' \code{{directory_path}/{folder_name}_grades.csv}.
#'
#' @returns Called for its side effects. Returns \code{NULL} invisibly.
main <- function() {
  all_entries <- list.dirs(directory_path, full.names = TRUE, recursive = FALSE)
  subdirs     <- all_entries[file.info(all_entries)$isdir %in% TRUE]

  if (length(subdirs) == 0L) {
    stop("No student subfolders found in ", directory_path)
  }

  message("Starting reliability test: ", N, " runs x ",
          length(subdirs), " students\n")

  for (folder in subdirs) {
    student_file <- file.path(folder,
                              stringr::str_glue("lab-{LAB_NUMBER}.qmd"))
    if (!file.exists(student_file)) {
      message("Skipping ", basename(folder), " — no lab-", LAB_NUMBER, ".qmd")
      next
    }

    folder_name  <- basename(folder)       # e.g. "lab-9_student_high"
    student_name <- stringr::str_remove(
      folder_name,
      stringr::str_glue("(?i)^lab-{LAB_NUMBER}_")
    )
    output_csv   <- file.path(directory_path,
                              paste0(folder_name, "_grades.csv"))

    message("Grading ", student_name, " (", N, " runs) ...")
    df <- grade_n_times(student_file, student_name, N)
    readr::write_csv(df, output_csv, na = "")
    message("  Saved: ", output_csv, "\n")
  }

  message("Reliability test complete.")
}

# run
if (identical(environment(), globalenv())) {
  tryCatch(main(), error = function(e) {
    message("Error: ", conditionMessage(e))
    if (!interactive()) quit(save = "no", status = 1)
  })
}

File: Python/reliability_test.py
"""Reliability test for the Chat Completions grading pipeline (Python).

Runs ``grade_student_qmd()`` N times for every student submission and
writes one CSV per student with one row per run. Re-running appends to
existing CSVs with continuous run numbering, so 10 invocations of N=10
accumulate to 100 rows per student.

Output files are written beside the student submission folders::

    {BASE_DIR}/{folder_name}_grades.csv
    e.g. "lab-9_student_high_grades.csv"

Columns: ``Run``, ``Total``, ``OverallComment``,
``Q1``–``Q{Q_COUNT}``, ``Q1_feedback``–``Q{Q_COUNT}_feedback``.

Usage::

    python Python/reliability_test.py                    # lab 9, 10 runs (defaults)
    python Python/reliability_test.py --lab-number 4     # lab 4, 10 runs
    python Python/reliability_test.py -L 4 --n 25        # lab 4, 25 runs
"""

import argparse
import csv
import os
from pathlib import Path

import grading_context
from grade_student import grade_student_qmd

Q_COLS          = [f"Q{i}" for i in range(1, grading_context.Q_COUNT + 1)]
Q_FEEDBACK_COLS = [f"Q{i}_feedback" for i in range(1, grading_context.Q_COUNT + 1)]
FIELDNAMES      = ["Run", "Total", "OverallComment"] + Q_COLS + Q_FEEDBACK_COLS


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _get_run_offset(csv_path: Path) -> int:
    """Return the highest Run number already recorded in *csv_path*, or 0.

    Used to derive the starting offset for a new batch of runs so that
    run numbers continue from where a previous invocation left off.

    :param csv_path: Path to an existing per-student grades CSV.
    :type csv_path: pathlib.Path
    :returns: Maximum value of the ``Run`` column, or ``0`` if the file
        does not exist or contains no valid run numbers.
    :rtype: int
    """
    if not csv_path.exists():
        return 0
    with csv_path.open(newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        runs = []
        for row in reader:
            try:
                runs.append(int(row["Run"]))
            except (KeyError, ValueError):
                pass
    return max(runs, default=0)


def grade_n_times(student_path: Path, n_runs: int, run_offset: int) -> list[dict]:
    """Grade one student submission *n_runs* times and return a list of row dicts.

    Each dict contains a ``Run`` key (offset-adjusted) plus all grade and
    feedback fields. Per-run exceptions are caught and stored as error rows
    so a single API failure does not abort the student's batch.

    :param student_path: Path to the student's ``.qmd`` file.
    :type student_path: pathlib.Path
    :param n_runs: Number of grading runs to perform.
    :type n_runs: int
    :param run_offset: Added to each run index so numbering continues from
        the last row of an existing CSV.
    :type run_offset: int
    :returns: List of *n_runs* row dicts ready for :class:`csv.DictWriter`.
    :rtype: list[dict]
    """
    rows = []
    for i in range(1, n_runs + 1):
        run_number = run_offset + i
        print(f"  Run {run_number} ({i}/{n_runs}) ...")

        try:
            result    = grade_student_qmd(student_path)
            questions = result.get("questions", {})
            row = {
                "Run":            run_number,
                "Total":          result.get("total"),
                "OverallComment": result.get("overall_comment", ""),
            }
            for q in Q_COLS:
                qinfo         = questions.get(q, {})
                row[q]        = qinfo.get("grade")
                row[f"{q}_feedback"] = qinfo.get("feedback")

        except Exception as e:
            print(f"    ERROR on run {run_number}: {e}")
            row = {"Run": run_number, "Total": None,
                   "OverallComment": f"Error: {e}"}
            for q in Q_COLS:
                row[q]               = None
                row[f"{q}_feedback"] = None

        rows.append(row)
    return rows


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main(n_runs: int = 10, lab_number: int = 9) -> None:
    """Run the reliability test across all student submissions.

    Calls :func:`grading_context.configure` to set lab-specific paths, then
    walks the lab submission folder for student submission files, grades each
    student :data:`n_runs` times, and writes (or appends to) a per-student
    CSV beside the submission folder.

    :param n_runs: Number of grading runs per student per invocation.
    :type n_runs: int
    :param lab_number: Lab number to test.
    :type lab_number: int
    :returns: None
    :rtype: None
    :raises FileNotFoundError: If the lab submission folder does not exist.
    """
    grading_context.configure(lab_number)
    base_dir = grading_context.BASE_LAB_DIR / f"lab-{lab_number}"

    student_paths = sorted(base_dir.rglob(f"lab-{lab_number}.qmd"))
    if not student_paths:
        raise FileNotFoundError(
            f"No lab-{lab_number}.qmd files found under {base_dir}"
        )

    print(f"Starting reliability test: {n_runs} runs x "
          f"{len(student_paths)} students\n")

    for path in student_paths:
        folder_name = path.parent.name          # e.g. "lab-9_student_high"
        output_csv  = base_dir / f"{folder_name}_grades.csv"

        run_offset  = _get_run_offset(output_csv)
        start, end  = run_offset + 1, run_offset + n_runs
        print(f"Grading {folder_name} (runs {start}–{end}) ...")

        rows = grade_n_times(path, n_runs, run_offset)

        write_header = not output_csv.exists()
        with output_csv.open("a", newline="", encoding="utf-8") as f:
            writer = csv.DictWriter(f, fieldnames=FIELDNAMES)
            if write_header:
                writer.writeheader()
            writer.writerows(rows)

        print(f"  Saved: {output_csv}\n")

    print("Reliability test complete.")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Run the grading pipeline N times per student."
    )
    parser.add_argument(
        "--n", type=int, default=10,
        help="Number of grading runs per student (default: 10)"
    )
    parser.add_argument(
        "--lab-number", "-L",
        type=int,
        default=int(os.getenv("LAB_NUMBER", 9)),
        help="Lab number to test (default: %(default)s)",
    )
    args = parser.parse_args()
    main(n_runs=args.n, lab_number=args.lab_number)
